File: src/08.dbpointer/05.16.ts
export function threeSumClosest(nums: number[], target: number): number {
    nums.sort((a, b) => a - b);

    let ans = nums[0] + nums[1] + nums[2];

    for (let i = 0; i < nums.length; i++) {
        let left = i + 1;
        let right = nums.length - 1;

        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];

            if (Math.abs(target - sum) < Math.abs(target - ans)) {
                ans = sum;
            }

            if (sum < target) {
                left++;
            } else if (sum > target) {
                right--;
            } else {
                return ans;
            }
        }
    }

    return ans;
}

File: src/08.dbpointer/05.16.go
package dbpointer

import (
	"math"
	"sort"
)

func ThreeSumClosest(nums []int, target int) int {
	sort.Ints(nums)

	ans := nums[0] + nums[1] + nums[2]

	for i := 0; i < len(nums); i++ {
		l := i + 1
		r := len(nums) - 1

		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			if math.Abs(float64(target-sum)) < math.Abs(float64(target-ans)) {
				ans = sum
			}

			if sum < target {
				l++
			} else if sum > target {
				r--
			} else {
				return ans
			}
		}
	}

	return ans
}
